### ISSUE TITLE
Proposal : Timecode to CV

### ISSUE THREAD

**Issue by RemiCollin** [VCVRack/Rack#1669]
### Summary

The goal of this proposal is to brainstorm a way to add "DAW-like" global transport functionality to VCV rack. 
With the upcoming of "Rack for Daws", it's likely that lot of users will expect VCV's sequencers and other time-based module to play "in sync" with their DAW, regardless of the position from which the playback is started in a track. 

Furthermore, having such a module in VCV standalone would make it easier to develop modules for DAW like functions as automation recorders, WAV players, etc..

### Proposed design

One approach could be to have a "Transport module" outputs two CVs :

- One CV output would indicate the index of the beat which is currently playing. A function would allow to convert a 0-10v signal to an absolute beat index. The max beat index would be ridiculously high to face any situation.

- One Phase CV would output the position inside the current beat with a  0-10v signal, like ZZC Clock already does : https://zzc.github.io/en/clock-manipulation/clock

**Comment by RichieHindle** [comment]
Thanks for starting this discussion!  A few notes and questions (some of which might be rather naive!)

VCV-Host (which I co-wrote and co-maintain) already sends a transport position to the VST plugin using the VST2 standard.  This sends both a beat position (as a double, so it can be fractional) and a sample position.  Currently, the beat position and sample position just grow indefinitely (unless reset by an explicit Stop or Reset transport command).  It also sends a tempo, which (if there's a clock wired in) is derived from the duration of the gaps between clock pulses.

All that should work well with a timecode input.  It's possible that a tempo CV might be useful, here and elsewhere, but Host will be happy to derive the  tempo from the gaps between beats, as it does now.

For anything with a clock input (eg. a sequencer or tempo-synced LFO) would a timecode input simply override the clock input?

You mention ZZC, in terms of its phase standard, which seems sensible.  ZZC also suggests a standard for tempo of one volt per beat per second.  That seems a bit limiting, and is different from the [VCV standard](https://vcvrack.com/manual/VoltageStandards) of `f = f0⋅2^V` where `f0` is 2Hz  (120bpm).  If a tempo standard is needed, I vote for the VCV standard.

One wire would be better than two - a standard for sending these CVs down the channels of a polyphonic wire would be good: channel 0 is the beat index, channel 1 is the phase, channel 2 is the tempo (if that becomes a thing), etc.

I wonder how the timecode gets sent to sequencers that don't start at the beginning of the piece?  Let's say you have a 32-bar piece in 4/4 time, and you have a sequencer containing an 8-bar sequence (32 beats) that you want to start playing at bar 12 (beat 48).  Currently you might use Janneker to send a start pulse after 48 beats.  How does this work with a timecode?  Does the sequencer receive the global timecode and you have to tell it to come in at beat 48 (which would be a feature that all sequencers would need to support)?  Or do you use a timeline module to mediate between the global timecode and the sequencer, only sending the timecode between beats 48 and 80, and subtracting 48 from it?

**Comment by AndrewBelt** [comment]
I think for the purpose of this proposal, only beat position is needed, not an actual "time"code (although that would be a bonus). Beat position can simply be a 24[PPQN](https://en.wikipedia.org/wiki/Pulses_per_quarter_note) (or 48PPQN) counter, meaning 24*4=96 (or 48*4=192) counts per measure. If we could simply send 32-bit integers over cables, we could simply send the position, but of course we're limited to float32. Here are our options.
- Cast integers to float. Disadvantages are that you can only encode ~24 bit integers, which would be 24 hours at 120bpm or something. Also, you'd be throwing around millions of virtual volts, which can break things when you "incorrectly" patch, which is what makes modular fun.
- Use the "low mantissa bits" of float32, so take your int32 and multiply by `std::pow(2.f, -23)` or whatever the minimum granularity is. Disadvantage is that conversion to analog voltage (with an audio interface) and back would destroy the signal due to precision.
- Use a digital (0V and 10V binary) timecode signal that actually works with analog conversion. Disadvantages are that it takes more CPU and a bit of latency to encode and decode. I don't want to write the encoder/decoder, but if someone would like to, I can add an implementation to the `dsp::` namespace.

**Comment by RemiCollin** [comment]
> You mention ZZC, in terms of its phase standard, which seems sensible. ZZC also suggests a standard for tempo of one volt per beat per second. That seems a bit limiting, and is different from the [VCV standard](https://vcvrack.com/manual/VoltageStandards) of `f = f0⋅2^V` where `f0` is 2Hz (120bpm). If a tempo standard is needed, I vote for the VCV standard.

Those are 2 different things. I agree that we should comply to the VCV standard for Tempo CV, but phase is more than that; I think it's an efficient way of carrying both tempo information & positioning at the same time. 
When using a trig input, you needs two trigs to calculate the actual tempo. With a phase input, you only need 2 samples, because you can deduce the tempo by calculating the phase difference between two samples. Therefore it's much more reliable to sync modules that do clock modulation, like a synced lfo for example. 
It's not only used by the ZZC modules; it's used in Bitwig Studio "The grid", Max/MSP and probably other virtual modular environments.  

> I wonder how the timecode gets sent to sequencers that don't start at the beginning of the piece? Let's say you have a 32-bar piece in 4/4 time, and you have a sequencer containing an 8-bar sequence (32 beats) that you want to start playing at bar 12 (beat 48). Currently you might use Janneker to send a start pulse after 48 beats. How does this work with a timecode? Does the sequencer receive the global timecode and you have to tell it to come in at beat 48 (which would be a feature that all sequencers would need to support)? Or do you use a timeline module to mediate between the global timecode and the sequencer, only sending the timecode between beats 48 and 80, and subtracting 48 from it?

That would be a workaround in that situation I guess. But we must keep in mind, that there is no way it can be perfect. One patch could send random triggers to start sequencer at any point and a timecode won't help with that.

> I think for the purpose of this proposal, only beat position is needed, not an actual "time"code (although that would be a bonus). Beat position can simply be a 24[PPQN](https://en.wikipedia.org/wiki/Pulses_per_quarter_note) (or 48PPQN) counter, meaning `24*4=96` (or `48*4=192`) counts per measure.

Didn't think about using PPQN. Indeed it could be a good middle ground. I wanted to mention phase as It's the most precise synchronisation method I'm aware of, plus as I wrote earlier it's widely used in virtual modular environments, and proposing it in an official VCV module would encourage plugin developers to adopt it, I guess. 

> * Use the "low mantissa bits" of float32, so take your int32 and multiply by `std::pow(2.f, -24)`. Disadvantage is that conversion to analog voltage (with an audio interface) and back would destroy the signal due to precision.

Can you think of a situation where this would be used in hardware ?

**Comment by RichieHindle** [comment]
(Sorry, I didn't make myself clear.  I wasn't recommending against the ZZC definition of phase; I was only recommending against their definition of tempo.)

**Comment by AndrewBelt** [comment]
Signals in VCV Rack need to be immune to small timing or amplitude changes. Rigorously, a timing deviation applied to a signal `f(t) -> f(t + h)` or amplitude deviation `f(t) -> f(t) + h` should have a bounded effect on a signal. In other words, the timecode standard should not be [chaotic](https://en.wikipedia.org/wiki/Chaos_theory). Why? Because users often do things to signals that unnoticeably change its quality, e.g. DAC/ADC conversion, recording to a 24-bit WAV and playing it back, attenuating it by 0.999 or 1.001, offsetting it by 0.001V, delaying the signal by a dozen samples, adding a -60dB noise floor, adding slight harmonic distortion, etc.

Your "proposed design" of using two CVs for timecode is susceptable to timing differences between the two signals. If a module "rolls over" from beat 42 phase 0.9999 to beat 43 phase 0.0001, and the phase signal arrives 1 sample later than the beat signal, it will appear as beat 43 phase 0.9999, which is way off from the actual position. So this won't work.

But sending an integer count every 24th or 48th of a beat solves this issue, because nobody needs real-valued playhead position. All we need is a way to send an int32 "message" over CV using some method of sending bytes over a wire.

Or we could go with my first bullet point and use unbounded real numbers, which is how VST2 does it. The `struct VstTimeInfo` has a `double samplePos` in seconds and `double ppqPos` in quarter notes (beats).

**Comment by RemiCollin** [comment]
> Your "proposed design" of using two CVs for timecode is susceptable to timing differences between the two signals. If a module "rolls over" from beat 42 phase 0.9999 to beat 43 phase 0.0001, and the phase signal arrives 1 sample later than the beat signal, it will appear as beat 43 phase 0.9999, which is way off from the actual position. So this won't work.

That's a good point, in that context it's true the sample precision of the phase approach would be indeed an issue. 

> * Use a digital (0V and 10V binary) timecode signal that actually works with analog conversion. Disadvantages are that it takes more CPU and a bit of latency to encode and decode. I don't want to write the encoder/decoder, but if someone would like to, I can add an implementation to the `dsp::` namespace.

I don't realize how much latency this would add, around 32 samples if we want to encode 32 bits right ? I would personally be in favor of any option with no added latency; in a scenario as Richie suggested where one module would mutate the timecode to act as a timeline for other sequencers, this latency would add up between modules.

> * Cast integers to float. Disadvantages are that you can only encode ~24 bit integers, which would be 24 hours at 120bpm or something. Also, you'd be throwing around millions of virtual volts, which can break things when you "incorrectly" patch, which is what makes modular fun.

I think I like this one better because it's simple.

Another possible option I didn't think in the first place, would be to directly pass the timecode through `process(args)` to the modules, in the same way samplerate is communicated to plugins. This would, of course, require to add a global transport to VCV, which I'm not sure would fit really well in the current rack philosophy (but kind of make sense in the "Rack for Daw" context). Advantage is it would add no CPU overhead for encoding/decoding/clock following, and increase rack performance overall. Drawback is it would be less flexible to users (yet I can imagine workaround to it would exists). 

Not favoring one way or another,  just wanna lay every option on the table for collective brainstorming.

**Comment by AndrewBelt** [comment]
>around 32 samples if we want to encode 32 bits right ? 

Yeah, at least 32 bits, maybe ~40 or so depending on the bit protocol used. That's almost 1ms latency.

>I think I like this one better because it's simple.

I agree, but it has the obvious problem of generating millions of volts.

What if we use a standard of 1mV/beat? For example, if your DAW playhead is exactly at measure 42 (41 with zero-indexing) at 4/4 time signature, the signal would be `41*4*1mv = 0.164V`. This standard would allow for 83 minutes of playback at 120bpm while being below the standard 10V maximum voltage. If more than 10,000 beats are needed, the signal would simply go higher than 10V, but this is fine because users know that going to extremes (e.g with project length) will push other things to the extremes (e.g. like voltages).
- Good: Users can do weird things to the signals to get expected results. E.g. attenuating the timecode signal by 0.5 would slow down the tempo by half.
- Bad: Even a well-calibrated DAC/ADC cannot accurately represent <1mV details. E.g. if your audio interface is off by 0.1mV, your timing will be off by 10% of a beat, which is super noticeable. So this standard is mostly useless for hardware or hybrid hardware/software setups. However, there are better standards for hardware, such as MIDI, SMPTE, etc that could be converted to/from the 1mV/beat standard at the boundary between software and hardware.

**Comment by ptrumpet** [comment]
How about encoding it as non-signalling NaN.  you have 22 bits to play with for a 32 bit float.  was planning to attempt this for sending midi signals with bottom 8 bits midi control, and upper 14 bits data (7 lo + 7 hi).  if this was an approved method for signalling digital signals, an alternative would be to set up a protocol to send NaN with a few bits dedicated to the application (e.g. 6 bits protocol + 16 bits of data).  midi could also be sent as bytes and decoded.

other alternatives would be to provide a digital channel carried by the same cable.

**Comment by AndrewBelt** [comment]
@ptrumpet In Rack v2, if the engine detects that a module sends a non-finite number, it replaces it with 0. Or, it will disable the module, but I haven't decided yet.

**Comment by RichieHindle** [comment]
I like the 1mV/beat proposal.  In common cases it keeps the numbers sane, but it does extend beyond the common range.  You get to use friendly voltages and still get a high resolution for your timing.  And it doesn't introduce latency.

If 83 minutes seems a little short, we could use 1mV/bar, giving 333 minutes per 10V for 4/4 at 120bpm.

I don't know enough about how people might want to integrate hardware with this stuff to comment on that.  Something simple like using an external knob to scrub through a piece should work well - you don't need millisecond precision for that.

**Comment by AndrewBelt** [comment]
>we could use 1mV/bar

I believe beat count is invariant under changing your DAW's time signature. Measures are defined by `beat / beats per measure`.

For actual "time"codes, I propose 1mV/second.

**Comment by RemiCollin** [comment]
> If 83 minutes seems a little short, we could use 1mV/bar, giving 333 minutes per 10V for 4/4 at 120bpm.

If I understood @AndrewBelt 's idea, the signal could go over 10V so the theorical length while keeping 1mv precision would be much larger than 83 minutes @ 120bpm, is that correct ?

Speaking of hardware, I guess we can safely assume this won't be used outside of the digital domain for obvious reasons, and already existing better alternatives.

**Comment by AndrewBelt** [comment]
>the theorical length while keeping 1mv precision would be much larger than 83 minutes @ 120bpm, is that correct ?

Yes, the theoretical limit is essentially infinite. Several days before you start losing noticeable precision.

>Speaking of hardware, I guess we can safely assume this won't be used outside of the digital domain for obvious reasons, and already existing better alternatives.

Yes, that is a good summary.

**Comment by RemiCollin** [comment]
One more details if we go this route, I think `0,000V` should be equals to `position = -1` to allow modules to detect start when `timecode = 0,001V`. I'm not really aware of how the VST SDK works, but I guess DAW pre-roll is not communicated to VST plugins, right ? Just so we can discard the necessity of using negative values.

**Comment by AndrewBelt** [comment]
0V should definitely correspond to the origin position (start of the first measure).

**Comment by RemiCollin** [comment]
So we should rely on another trigger for start/play detection ?

**Comment by squinkylabs** [comment]
Dead horse, by now, but one number for position is way better than two! For the reason sited.

**Comment by Coirt** [comment]
An LFO will gracefully scroll trough any timeline, especially a saw. Add user defined sample and hold and you're laughing!
